### PR TITLE
Fixed CompositeID Marshaling and expanded test case to cover

### PR DESCRIPTION
**Pull request by tim-speed**
Relates: manifoldco/engineering#6573

Thanks to James for helping me realize what was wrong.
Never use a pointer signature for implementing a Marshaler!

### CODE REVIEW

**Comment by tim-speed** [comment]
Coverage is still 100%

**Comment by tim-speed** [comment]
> Edit (notes from James): an ID should probably not need all the extra information like the domain. Maybe ID should be a different reference type which we use in the Composite IDs.

This is why I added a ManifoldID type, it's to facilitate the conversion, it adheres to the interface of CompositeID but still has the underlying footprint of an ID.

> I feel like we can refactor the the ManifoldID into the regular ID object. This way, end users don't have to worry about which ID to use and the refactoring is abstracted in the ID item.

For this to work I would have to extend the Unmarshalling of FlexID to support the traditional ID JSON String. This could work, but it could be confusing when looking at the data if its not in a consistent format.

The simplest solution would be to do away with the notion of CompositeID as an interface, but instead make it a struct that is FlexID, also getting rid of ManifoldID and just having the FromID function to create a FlexID from an ID; but I don't like that because the whole point of this is to allow for support of external identifiers ( also kind of why I had this in it's own `id` package ), so if we ever need to add in representations of IDs in other systems it would be more structured, we could just create a new type from their type implementing the CompositeID interface, but you could just straight up build a FlexID too, so it doesn't really matter functionally, it was more about structure; I was thinking that I'd prefer to pass around an interface when possible, and store a concrete type `FlexID`, it's all preference I think.

@manifoldco/gophers anyone have anything to add? This is blocking a lot so we should get a resolution asap.

File: flexid.go [hunk 1]
```diff
@@ -118,44 +117,41 @@ type CompositeID interface {
...
+func (m ManifoldID) String() string {
 	return fmt.Sprintf("%s%s%s%s%s", m.Domain(), pathSeperator, m.Type(),
 		pathSeperator, m.ID())
 }
 
 // Validate implements the Validate interface for goswagger
-func (m *ManifoldID) Validate(_ strfmt.Registry) error {
-	return manifold.ID(*m).Validate(nil)
+func (m ManifoldID) Validate(_ strfmt.Registry) error {
+	return ID(m).Validate(nil)
```
[the remark this answers is not in the document]
Comment (tim-speed) [reply flexid.go]: sure

**Comment by enmand** [comment]
I like @jbowes' and @jelmersnoeck's suggestions. If we kept the CompositeID `interface{}`, we don't need the domain part on `ID` (since we know it's our own ID), but we could make it fit the `CompositeID`. interface, then we could pass around `ID` or `FlexID` as an `CompositeID`.

 The `FlexID` is tripping me up a bit (I think it would be better as a struct), with a `NewFlexID(domain, type, id string) FlexID` func to create one.

Maybe we don't need an `AsFlexID` in the `CompositeID` (unless we had an `AsID` too, and for whatever ID types might come later) -- we could just type-cast the interface and check the conversion to know the type or switch on the type.

I think if we ditched `ManifoldID` and made `ID` work as a `CompositeID` along with `FlexID`, it might suite our needs (and maybe not need a conversion of IDs in our DB?)

**Comment by tim-speed** [comment]
> I like @jbowes' and @jelmersnoeck's suggestions. If we kept the CompositeID interface{}, we don't need the domain part on ID (since we know it's our own ID), but we could make it fit the CompositeID. interface, then we could pass around ID or FlexID as an CompositeID.

The main issue with making ID compliant is that it will not convert correctly to JSON, the goal of this is to introduce a new identifier type that can be used alongside the existing ID type, not replace it. That new type is the FlexID, which has the ManifoldID to help convert existing IDs only as needed.

> The FlexID is tripping me up a bit (I think it would be better as a struct), with a NewFlexID(domain, type, id string) FlexID func to create one.

A constructor would make sense there that does validation on the inputs, I can definitely add that, right now it's just setup to store the IDs, FlexID is the interchangeable data form of a CompositeID

> Maybe we don't need an AsFlexID in the CompositeID (unless we had an AsID too, and for whatever ID types might come later) -- we could just type-cast the interface and check the conversion to know the type or switch on the type.

AsManifoldID specificly exists only on the FlexID type, for convenience and errors if the FlexID can not be converted. AsID exists only on the ManifoldID type for convenience, because the conversion was bloating the code in Marketplace. The Idea is that everything will be stored as a FlexID, a series of three string parts, `Domain, Type, ID` and that everything will be either converted from or two this format. 

> I think if we ditched ManifoldID and made ID work as a CompositeID along with FlexID, it might suite our needs (and maybe not need a conversion of IDs in our DB?)

*See first comment*

**Pull request merged**